### PR TITLE
Issue 233 add simplifications to solver

### LINKED ISSUE #233: apply model simplifications to solver
re. https://github.com/pybamm-team/PyBaMM/pull/228#pullrequestreview-220002596

have unary and binary simplifications in place (see #228 ), need to apply these in the solver before the simulation loop.

note, have focused on testing scalar simplifications only at this point, so expecting problems with vectors.

### PR DESCRIPTION
**Pull request by martinjrobins**
# Description

- add more simplifications to expression tree
   1. given a term in the expression tree consisting of multiplications/divisions, pybamm will attempt to simplify it by grouping together those sub-terms that are constant, and return a value, and those that are not. Those that are constant are reduced to a single sub-term
   2. given a term in the expression tree consisting of multiplications/divisions/matrix-multiplications, pybamm will attempt to simplify it by grouping together neighbouring sub-terms that are constant, and return a value. Note that this is different to 1, due to the presence of matrix-multiplications pybamm does not attempt to reorder the terms.
  3. given a term in the expression tree consisting of additions/subtractions, pybamm will attempt to simplify it by grouping together those sub-terms that are constant, and return a value, and those that are not. Those that are constant are reduced to a single sub-term

- adds simplifications to ode and dae solvers

Fixes #233 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pybamm/expression_tree/binary_operators.py [hunk 1]
```diff
@@ -7,6 +7,282 @@
...
+
+            if child == left_child and this_class == pybamm.Division:
+                in_numerator = not in_numerator
+
+    flatten(None, myclass, left, right, True)
+
+    # check if there is a matrix multiply in the numerator (if so we can't reorder it)
+    has_matrix_multiply = myclass == pybamm.MatrixMultiplication
+    for t in numerator_types:
+        has_matrix_multiply |= t == pybamm.MatrixMultiplication
```
Comment (tinosulzer) [inline pybamm/expression_tree/binary_operators.py]:
```python3
has_matrix_multiply = (myclass == pybamm.MatrixMultiplication or any([t == pybamm.MatrixMultiplication for t in numerator_types]))
```
might be clearer?

File: pybamm/expression_tree/binary_operators.py [hunk 2]
```diff
@@ -7,6 +7,282 @@
...
+        for child in source:
+            if child.is_constant():
+                result = child.evaluate_ignoring_errors()
+                if result is not None:
+                    constant.append(child)
+                else:
+                    nonconstant.append(child)
+            else:
+                nonconstant.append(child)
+        return constant, nonconstant
```
Comment (tinosulzer) [inline pybamm/expression_tree/binary_operators.py]: see suggestion above

File: pybamm/expression_tree/binary_operators.py [hunk 3]
```diff
@@ -7,6 +7,282 @@
...
+
+    """
+    numerator = []
+    numerator_types = []
+
+    # recursive function to flatten a term involving only additions or subtractions
+    def flatten(previous_class, this_class, left_child, right_child):
+        for child in [left_child, right_child]:
+            if isinstance(child, pybamm.Addition) \
+                    or isinstance(child, pybamm.Subtraction):
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pybamm/expression_tree/binary_operators.py]: Yea, agree that this code is quite convoluted, it was a bit tricky getting it to work. I'll try and expand on the docstring to give some examples

>  Would combinations of multiplications and additions like
> 
> ```python
> (1 + (2*3 + 1 + a))/(4*2 + 5)
> ```
> 
> get simplified appropriately?

They would, partially, but this code would only be involved for a bit of it. Remember that the simplify occurs recursively, so the 2*3, 4*2 would be first simplified to 6 and 8, then the 8 + 5 would be simplified, then the 6 + 1, then finally this new code would take over to simplify (1 + (7 + a)) to 8 + a, and then (8 + a)/13 would **not** be simplified. This latter bit would require some more code to expand any multiplies or divides containing additions/subtractions, should do another issue for this

**Comment by martinjrobins** [comment]
Yea, agree that this code is quite convoluted, it was a bit tricky getting it to work. I'll try and expand on the docstring to give some examples

>  Would combinations of multiplications and additions like
> 
> ```python
> (1 + (2*3 + 1 + a))/(4*2 + 5)
> ```
> 
> get simplified appropriately?

They would, partially, but this code would only be involved for a bit of it. Remember that the simplify occurs recursively, so the 2*3, 4*2 would be first simplified to 6 and 8, then the 8 + 5 would be simplified, then the 6 + 1, then finally this new code would take over to simplify (1 + (7 + a)) to 8 + a, and then (8 + a)/13 would **not** be simplified. This latter bit would require some more code to expand any multiplies or divides containing additions/subtractions, should do another issue for this

File: pybamm/expression_tree/binary_operators.py [hunk 4]
```diff
@@ -7,6 +7,282 @@
...
+                left = copy.deepcopy(child.children[0])
+                left.parent = None
+                right = copy.deepcopy(child.children[1])
+                right .parent = None
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pybamm/expression_tree/binary_operators.py]: ahh yes, forgot about that. cheers!

File: pybamm/expression_tree/binary_operators.py [hunk 5]
```diff
@@ -7,6 +7,282 @@
...
+                    numerator_types.append(this_class)
+
+    flatten(None, myclass, left, right)
+
+    # function to partition a source list of symbols into those that return a constant
+    # value, and those that do not
+    def partition_by_constant(source, types):
+        constant = []
+        nonconstant = []
+        constant_t = []
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pybamm/expression_tree/binary_operators.py]: yea, sorry, that is a C++ convention, I'll make it more obvious

**Comment by martinjrobins** [comment]
excellent. Glad to hear you get the speed gains!

**Pull request merged**